src/thunderbots/software/ai/hl/stp/evaluation/indirect_chip.h

```diff
@@ -0,0 +1,121 @@
+ */
+namespace Evaluation
+{
+    typedef Point Vector2;
+    template <size_t N>
+    using Poly     = std::array<Vector2, N>;
+    using Triangle = Poly<3>;
+
+    /**
+     * Returns the target point that the chipper will shoot at and the chaser will meet
+     * ball at. Given the enemy team robot's positions, filters and creates a vector of
+     * triangles to be used as a parameter. The target is where ball will land according
+     * to chipping calibration.
+     *
+     * @param World Object
+     *
+     * @return Point to chip and chase at; If empty, point is not valid
+     */
+    std::optional<Point> indirect_chip_and_chase_target(const World& world);
+
+    /**
+     * Returns the target point that the chipper will shoot at and the chaser will meet
+     * ball at. Given the filtered vector of triangles, determines if the triangles are
+     * empty and if the largest triangle is within reach. The target is where ball will
+     * land according to chipping calibration.
+     *
+     * @param triangles A vector of triangles that is already filtered
+     * @param ball_position Position of the ball
+     *
+     * @return Point to chip and chase at; If empty, point is not valid
+     */
+    std::optional<Point> indirect_chip_and_chase_target(
+        const std::vector<Triangle>& triangles, Point ball_position);
+
+    /**
+     * Creates a vector of triangles that picks all permutations of points from the
+     * list of all non-goalie enemy players as well as the four points returned by
+     * get_chip_area_target_corners.
+     *
+     * @param World Object
+     * @param enemy_players
+     *
+     * @return Vector of triangles
+     */
+    std::vector<Triangle> get_all_triangles(const World& world,
+                                            std::vector<Point> enemy_players);
+
+    /**
+     * Given a vector of triangles, returns a new vector that only contains triangles
+     * without enemy robots.
+     *
+     * @param triangles
+     * @param enemy_players
+     *
+     * @return Vector of triangles with no enemy robots
+     */
+    std::vector<Triangle> filter_open_triangles(std::vector<Triangle> triangles,
+                                                std::vector<Point> enemy_players);
+
+    /**
+     * Returns the center point and area of the given triangle.
+     *
+     * @param triangle
+     *
+     * @return Centre point of triangle
+     * @return Area of triangle
+     */
+    std::pair<Point, double> get_triangle_center_and_area(Triangle triangle);
+
+    /**
+     * Remove all Triangles in a given list whose centers do not fall
+     * within the rectangle returned by get_chip_target_area.
+     *
+     * @param World Object
+     * @param triangles
+     *
+     * @return Valid triangles
+     */
+    std::vector<Triangle> remove_outofbounds_triangles(const World& world,
+                                                       std::vector<Triangle> triangles);
```

Comment:
optional; but it might be good to have this function take a `Rectangle` and rename it to `removeTrianglesOutsideRectangle`. This would make it de-coupled from the `get_chip_target_area` function, so it would be easier to write robust tests for it.

The reason to do something like this is because if we ever change `get_chip_target_area` in the future, it _could_ break the tests for this function as well, but if they're totally separate then the `removeTrianglesOutsideRectangle` tests would be fine.

Obviously there's no expectation for you to know this, but here's a great article that I'd suggest giving a quick read. Learning a bit of software engineering theory can be really helpful! https://www.geeksforgeeks.org/software-engineering-coupling-and-cohesion/

Reply:
Thanks for the read! `removeTrianglesOutsideRectangle` now takes a `Rectangle`, so like you mentioned (I think this is what you meant), this function no longer strictly depends on `findBestChipTargetArea`.

My rectangle implementation feels a little wonky since in `removeTrianglesOutsideRectangle` I basically just took the `Rectangle` input apart and made it into a vector of `Points` again like before. I was wondering if that's an okay way of doing it?

Reply:
What you've done here looks pretty reasonable to me on all counts. :+1: